File: sdk/typescript/src/resources.ts
import * as api from './operations/index.js';
import * as models from './models/index.js';
import type {NetworkLogOptions, PageOptions} from './options.js';
import type {CallOptions, Transport} from './transport.js';

export interface VmContext {transport: Transport; id: string}
class Resource {
  constructor(protected readonly context: (options: CallOptions) => Promise<VmContext>) {}
}

export class Files extends Resource {
  async read(path: string, options: CallOptions = {}): Promise<Uint8Array> {
    const {transport, id} = await this.context(options);
    return api.downloadVmFile(transport, {id, path}, options);
  }
  async write(path: string, data: Uint8Array, options: CallOptions = {}): Promise<models.UploadResponse> {
    const {transport, id} = await this.context(options);
    return api.uploadVmFile(transport, {id, path, body: data}, options);
  }
  async list(path = '/', options: CallOptions & {depth?: number} = {}): Promise<models.FileListResponse> {
    const {transport, id} = await this.context(options);
    return api.listVmFiles(transport, {...options, id, ...(path === '/' ? {} : {path})}, options);
  }
  async history(checkpoint: string, options: PageOptions = {}): Promise<models.ChangesResponse> {
    const {transport, id} = await this.context(options);
    return api.getVmChanges(transport, {...options, id, checkpoint}, options);
  }
}
export class Snapshots extends Resource {
  async list(options: CallOptions = {}): Promise<models.SnapshotsList> {
    const {transport, id} = await this.context(options);
    return api.listVmSnapshots(transport, {id}, options);
  }
  async status(options: CallOptions = {}): Promise<models.SnapshotsStatus> {
    const {transport, id} = await this.context(options);
    return api.getVmSnapshotsStatus(transport, {id}, options);
  }
}
export class Stats extends Resource {
  async summary(options: CallOptions = {}): Promise<models.VmStatsSummaryResponse> {
    const {transport, id} = await this.context(options);
    return api.getVmStatsSummary(transport, {id}, options);
  }
  async details(options: CallOptions = {}): Promise<models.VmStatsDetailResponse> {
    const {transport, id} = await this.context(options);
    return api.getVmStatsDetail(transport, {id}, options);
  }
}
export class Container extends Resource {
  async status(options: CallOptions = {}): Promise<models.ContainerStatusResponse> {
    const {transport, id} = await this.context(options);
    return api.getVmContainer(transport, {id}, options);
  }
}

export interface Port {
  id: string;
  guest: number;
  host: number | null;
  authenticate: boolean;
  url?: string;
  bootstrapToken?: string;
  expiresInSeconds?: number;
}

function port(exposure: models.ExposureInfo): Port {
  return {
    id: exposure.id, guest: exposure.guest_port, host: exposure.host_port ?? null,
    authenticate: exposure.access === models.ExposureAccess.HTTP_PREVIEW,
  };
}

export class Ports extends Resource {
  constructor(context: (options: CallOptions) => Promise<VmContext>,
              private readonly target: (options: CallOptions) => Promise<models.ExposureTarget>) {
    super(context);
  }
  async open(guest: number, options: CallOptions & {host?: number; authenticate?: boolean} = {}): Promise<Port> {
    if (!Number.isInteger(guest) || guest < 1 || guest > 65_535) throw new TypeError('Guest port must be between 1 and 65535');
    const host = options.host ?? 0, authenticate = options.authenticate ?? false;
    if (!Number.isInteger(host) || host < 0 || host > 65_535) throw new TypeError('Host port must be between 0 and 65535');
    if (authenticate && host !== 0) throw new TypeError('Authenticated ports cannot select a host port');
    const {transport, id} = await this.context(options);
    const exposure = await api.createVmExposure(transport, {id, body: {
      guest_port: guest, host_port: host, target: await this.target(options),
      access: authenticate ? models.ExposureAccess.HTTP_PREVIEW : models.ExposureAccess.LOOPBACK_TCP,
    }}, options);
    const opened = port(exposure);
    if (!authenticate) return opened;
    let session: models.PreviewSessionResponse;
    try {
      session = await api.createVmPreviewSession(transport, {id, exposure_id: exposure.id}, options);
    } catch (error) {
      // The caller never receives a Port to close. Remove the exposure without
      // the caller's signal (it may be the reason we failed) and report the
      // session failure.
      await api.deleteVmExposure(transport, {id, exposure_id: exposure.id}).catch(() => undefined);
      throw error;
    }
    return {...opened, host: null, url: session.url, bootstrapToken: session.bootstrap_token,
      expiresInSeconds: session.expires_in_seconds};
  }
  async list(options: CallOptions = {}): Promise<Port[]> {
    const {transport, id} = await this.context(options);
    const response = await api.listVmExposures(transport, {id}, options);
    return response.exposures.map(port);
  }
  async close(opened: Port, options: CallOptions = {}): Promise<models.VmActionResponse> {
    if (typeof opened !== 'object' || opened === null || typeof opened.id !== 'string') {
      throw new TypeError('Port must be an object returned by vm.ports');
    }
    const {transport, id} = await this.context(options);
    return api.deleteVmExposure(transport, {id, exposure_id: opened.id}, options);
  }
}

export class Networks {
  constructor(private readonly transport: Transport) {}
  async create(name: string, options: CallOptions = {}): Promise<models.NetworkInfo> {
    return api.createNetwork(this.transport, {body: {name}}, options);
  }
  async list(options: CallOptions = {}): Promise<models.NetworkInfo[]> {
    return (await api.listNetworks(this.transport, options)).networks;
  }
  async inspect(networkId: string, options: CallOptions = {}): Promise<models.NetworkInfo> {
    return api.getNetwork(this.transport, {id: networkId}, options);
  }
  async delete(network: models.NetworkInfo, options: CallOptions = {}): Promise<models.VmActionResponse> {
    return api.deleteNetwork(this.transport, {id: network.id}, options);
  }
  async logs(network: models.NetworkInfo, options: NetworkLogOptions = {}): Promise<models.NetworkLogsResponse> {
    return api.getNetworkLogs(this.transport, {...options, id: network.id}, options);
  }
}

export class VmNetworks extends Resource {
  async list(options: CallOptions = {}): Promise<models.NetworkInfo[]> {
    const {transport, id} = await this.context(options);
    const response = await api.listNetworks(transport, options);
    return response.networks.filter(network => network.members.some(member => member.vm_id === id));
  }
  async attach(network: models.NetworkInfo, options: CallOptions = {}): Promise<models.NetworkInfo> {
    const {transport, id} = await this.context(options);
    return api.attachNetworkMember(transport, {id: network.id, vm_id: id}, options);
  }
  async detach(network: models.NetworkInfo, options: CallOptions = {}): Promise<models.NetworkInfo> {
    const {transport, id} = await this.context(options);
    return api.detachNetworkMember(transport, {id: network.id, vm_id: id}, options);
  }
}

export class McpTools {
  constructor(private readonly transport: Transport, private readonly profileId: string,
              private readonly serverId: string) {}
  async list(options: CallOptions = {}): Promise<models.McpToolsListResponse> {
    return api.listProfileMcpTools(this.transport, {profile_id: this.profileId, server_id: this.serverId}, options);
  }
  async call(name: string, arguments_: models.Value, options: CallOptions = {}): Promise<models.Value> {
    return api.callProfileMcpTool(this.transport, {
      profile_id: this.profileId, server_id: this.serverId, tool_id: name, body: arguments_,
    }, options);
  }
}

export class ProfileMcpServer {
  readonly tools: McpTools;
  constructor(private readonly transport: Transport, private readonly profileId: string,
              readonly info: models.McpServerInfoResponse) {
    this.tools = new McpTools(transport, profileId, info.name);
  }
  async refresh(options: CallOptions = {}): Promise<models.McpRefreshResponse> {
    return api.refreshProfileMcpServer(this.transport, {
      profile_id: this.profileId, server_id: this.info.name,
    }, options);
  }
}

export class ProfileMcp {
  constructor(private readonly transport: Transport, private readonly profile: models.ProfileSummary) {}
  async info(options: CallOptions = {}): Promise<models.ProfileMcpInfoResponse> {
    return api.getProfileMcpInfo(this.transport, {profile_id: this.profile.id}, options);
  }
  async servers(options: CallOptions = {}): Promise<models.McpServersListResponse> {
    return api.listProfileMcpServers(this.transport, {profile_id: this.profile.id}, options);
  }
  async defaultPermission(options: CallOptions = {}): Promise<models.McpDefaultPermissionResponse> {
    return api.getProfileMcpDefault(this.transport, {profile_id: this.profile.id}, options);
  }
  async get(name: string, options: CallOptions = {}): Promise<ProfileMcpServer> {
    const matches = (await this.servers(options)).filter(server => server.name === name);
    if (matches.length !== 1) throw new TypeError(`Expected one MCP server named ${JSON.stringify(name)}, found ${matches.length}`);
    return new ProfileMcpServer(this.transport, this.profile.id, matches[0] as models.McpServerInfoResponse);
  }
}

export class Profiles {
  constructor(private readonly transport: Transport) {}
  async list(options: CallOptions = {}): Promise<models.ProfileSummary[]> {
    return (await api.listProfiles(this.transport, options)).profiles;
  }
  mcp(profile: models.ProfileSummary): ProfileMcp {
    if (typeof profile !== 'object' || profile === null || typeof profile.id !== 'string') {
      throw new TypeError('Profile must be an object returned by hypervisor.profiles');
    }
    return new ProfileMcp(this.transport, profile);
  }
}

File: sdk/rust/src/resources.rs
use crate::client::Client;

use crate::{models, operations as api, DiagnosticOptions, NetworkLogOptions, Result, TriageOptions, VM};

/// Paths are the guest's: `/root/x` in a VM, `/workspace/x` in its container,
/// or `x` relative to the workspace; see [`Files::exact`].
pub struct Files<'a>(pub(crate) &'a VM, pub(crate) bool);
pub struct Stats<'a>(pub(crate) &'a VM);
pub struct Container<'a>(pub(crate) &'a VM);
pub struct Ports<'a>(pub(crate) &'a VM);
pub struct VmNetworks<'a>(pub(crate) &'a VM);
pub struct Networks<'a>(pub(crate) &'a Client);
pub struct Debug<'a>(pub(crate) &'a Client);
pub struct Profiles<'a>(pub(crate) &'a Client);
pub struct ProfileMcp<'a> {
    client: &'a Client,
    profile_id: String,
}
pub struct ProfileMcpServer<'a> {
    client: &'a Client,
    profile_id: String,
    pub info: models::McpServerInfoResponse,
}
pub struct McpTools<'a> {
    client: &'a Client,
    profile_id: String,
    server_id: String,
}

impl Files<'_> {
    /// Take every path literally, relative to the workspace root, even when it
    /// is absolute: `/root/x` is then the workspace's `root/x`.
    #[must_use]
    pub const fn exact(self) -> Self {
        Self(self.0, true)
    }

    fn exact_param(&self) -> Option<bool> {
        self.1.then_some(true)
    }

    pub async fn read(&self, path: &str) -> Result<Vec<u8>> {
        let params = api::DownloadVmFileParams {
            id: self.0.resolve().await?,
            path: path.into(),
            exact: self.exact_param(),
        };
        api::download_vm_file(&self.0.client.transport, &params, self.0.client.options).await
    }

    pub async fn write(&self, path: &str, data: Vec<u8>) -> Result<models::UploadResponse> {
        let params = api::UploadVmFileParams {
            id: self.0.resolve().await?,
            path: path.into(),
            exact: self.exact_param(),
            body: data,
        };
        api::upload_vm_file(&self.0.client.transport, &params, self.0.client.options).await
    }

    /// An empty `path` lists the workspace root.
    pub async fn list(&self, path: &str, depth: Option<i64>) -> Result<models::FileListResponse> {
        let params = api::ListVmFilesParams {
            id: self.0.resolve().await?,
            path: (!path.is_empty()).then(|| path.to_owned()),
            depth,
            exact: self.exact_param(),
        };
        api::list_vm_files(&self.0.client.transport, &params, self.0.client.options).await
    }
}

impl Stats<'_> {
    pub async fn summary(&self) -> Result<models::VmStatsSummaryResponse> {
        api::get_vm_stats_summary(
            &self.0.client.transport,
            &api::GetVmStatsSummaryParams {
                id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await
    }

    pub async fn details(&self) -> Result<models::VmStatsDetailResponse> {
        api::get_vm_stats_detail(
            &self.0.client.transport,
            &api::GetVmStatsDetailParams {
                id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await
    }
}

impl Container<'_> {
    pub async fn status(&self) -> Result<models::ContainerStatusResponse> {
        api::get_vm_container(
            &self.0.client.transport,
            &api::GetVmContainerParams {
                id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct Port {
    pub id: String,
    pub guest: u16,
    pub host: Option<u16>,
    pub authenticate: bool,
    pub url: Option<String>,
    pub bootstrap_token: Option<String>,
    pub expires_in_seconds: Option<u16>,
}

/// The bootstrap token opens a browser session on the workload, so it is
/// redacted the way [`crate::Registry`] redacts its credentials.
impl std::fmt::Debug for Port {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter
            .debug_struct("Port")
            .field("id", &self.id)
            .field("guest", &self.guest)
            .field("host", &self.host)
            .field("authenticate", &self.authenticate)
            .field("url", &self.url)
            .field(
                "bootstrap_token",
                &if self.bootstrap_token.is_some() {
                    "<redacted>"
                } else {
                    "<none>"
                },
            )
            .field("expires_in_seconds", &self.expires_in_seconds)
            .finish()
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct PortOptions {
    pub host: u16,
    pub authenticate: bool,
}

impl Port {
    fn from_exposure(exposure: models::ExposureInfo) -> Self {
        Self {
            id: exposure.id,
            guest: exposure.guest_port,
            host: exposure.host_port,
            authenticate: exposure.access == models::ExposureAccess::HttpPreview,
            url: None,
            bootstrap_token: None,
            expires_in_seconds: None,
        }
    }
}

impl Ports<'_> {
    pub async fn open(&self, guest: u16) -> Result<Port> {
        self.open_with(guest, PortOptions::default()).await
    }

    pub async fn open_with(&self, guest: u16, options: PortOptions) -> Result<Port> {
        if guest == 0 {
            return Err(crate::Error::InvalidInput("guest port must be between 1 and 65535"));
        }
        if options.authenticate && options.host != 0 {
            return Err(crate::Error::InvalidInput(
                "authenticated ports cannot select a host port",
            ));
        }
        let id = self.0.resolve().await?;
        let exposure = api::create_vm_exposure(
            &self.0.client.transport,
            &api::CreateVmExposureParams {
                id: id.clone(),
                body: models::ExposureRequest {
                    guest_port: guest,
                    host_port: options.host,
                    target: self.0.port_target().await?,
                    access: if options.authenticate {
                        models::ExposureAccess::HttpPreview
                    } else {
                        models::ExposureAccess::LoopbackTcp
                    },
                },
            },
            self.0.client.options,
        )
        .await?;
        let mut port = Port::from_exposure(exposure);
        port.authenticate = options.authenticate;
        if options.authenticate {
            let session = api::create_vm_preview_session(
                &self.0.client.transport,
                &api::CreateVmPreviewSessionParams {
                    id: id.clone(),
                    exposure_id: port.id.clone(),
                },
                self.0.client.options,
            )
            .await;
            let session = match session {
                Ok(session) => session,
                Err(error) => {
                    // The caller never receives a Port to close, so remove the
                    // exposure best-effort and report the session failure.
                    let _ = api::delete_vm_exposure(
                        &self.0.client.transport,
                        &api::DeleteVmExposureParams {
                            id,
                            exposure_id: port.id,
                        },
                        self.0.client.options,
                    )
                    .await;
                    return Err(error);
                }
            };
            port.host = None;
            port.url = Some(session.url);
            port.bootstrap_token = Some(session.bootstrap_token);
            port.expires_in_seconds = Some(session.expires_in_seconds);
        }
        Ok(port)
    }

    pub async fn list(&self) -> Result<Vec<Port>> {
        let response = api::list_vm_exposures(
            &self.0.client.transport,
            &api::ListVmExposuresParams {
                id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await?;
        Ok(response.exposures.into_iter().map(Port::from_exposure).collect())
    }

    pub async fn close(&self, port: &Port) -> Result<models::VmActionResponse> {
        api::delete_vm_exposure(
            &self.0.client.transport,
            &api::DeleteVmExposureParams {
                id: self.0.resolve().await?,
                exposure_id: port.id.clone(),
            },
            self.0.client.options,
        )
        .await
    }
}

impl Networks<'_> {
    pub async fn create(&self, name: &str) -> Result<models::NetworkInfo> {
        api::create_network(
            &self.0.transport,
            &api::CreateNetworkParams {
                body: models::CreateNetworkRequest { name: name.into() },
            },
            self.0.options,
        )
        .await
    }

    pub async fn list(&self) -> Result<Vec<models::NetworkInfo>> {
        Ok(api::list_networks(&self.0.transport, self.0.options).await?.networks)
    }

    pub async fn inspect(&self, network_id: &str) -> Result<models::NetworkInfo> {
        api::get_network(
            &self.0.transport,
            &api::GetNetworkParams { id: network_id.into() },
            self.0.options,
        )
        .await
    }

    pub async fn delete(&self, network: &models::NetworkInfo) -> Result<models::VmActionResponse> {
        api::delete_network(
            &self.0.transport,
            &api::DeleteNetworkParams { id: network.id.clone() },
            self.0.options,
        )
        .await
    }

    pub async fn logs(
        &self,
        network: &models::NetworkInfo,
        options: NetworkLogOptions,
    ) -> Result<models::NetworkLogsResponse> {
        api::get_network_logs(
            &self.0.transport,
            &api::GetNetworkLogsParams {
                id: network.id.clone(),
                cursor: options.cursor,
                limit: options.limit,
                vm: options.vm,
                connection: options.connection,
                r#type: options.event_type,
                decision: options.decision,
                since: options.since,
                until: options.until,
            },
            self.0.options,
        )
        .await
    }
}

impl VmNetworks<'_> {
    pub async fn list(&self) -> Result<Vec<models::NetworkInfo>> {
        let vm_id = self.0.resolve().await?;
        let response = api::list_networks(&self.0.client.transport, self.0.client.options).await?;
        Ok(response
            .networks
            .into_iter()
            .filter(|network| network.members.iter().any(|member| member.vm_id == vm_id))
            .collect())
    }

    pub async fn attach(&self, network: &models::NetworkInfo) -> Result<models::NetworkInfo> {
        api::attach_network_member(
            &self.0.client.transport,
            &api::AttachNetworkMemberParams {
                id: network.id.clone(),
                vm_id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await
    }

    pub async fn detach(&self, network: &models::NetworkInfo) -> Result<models::NetworkInfo> {
        api::detach_network_member(
            &self.0.client.transport,
            &api::DetachNetworkMemberParams {
                id: network.id.clone(),
                vm_id: self.0.resolve().await?,
            },
            self.0.client.options,
        )
        .await
    }
}

impl Debug<'_> {
    pub async fn panics(&self, options: DiagnosticOptions) -> Result<models::PanicsResponse> {
        api::get_panics(
            &self.0.transport,
            &api::GetPanicsParams {
                since: options.since,
                limit: options.limit,
                id: None,
            },
            self.0.options,
        )
        .await
    }

    pub async fn triage(&self, options: TriageOptions) -> Result<models::TriageResponse> {
        api::get_triage(
            &self.0.transport,
            &api::GetTriageParams {
                since: options.since,
                limit: options.limit,
                id: options.vm_id,
            },
            self.0.options,
        )
        .await
    }
}

impl<'a> Profiles<'a> {
    pub async fn list(&self) -> Result<Vec<models::ProfileSummary>> {
        Ok(api::list_profiles(&self.0.transport, self.0.options).await?.profiles)
    }

    pub fn mcp(&self, profile: &models::ProfileSummary) -> ProfileMcp<'a> {
        ProfileMcp {
            client: self.0,
            profile_id: profile.id.clone(),
        }
    }
}

impl<'a> ProfileMcp<'a> {
    pub async fn info(&self) -> Result<models::ProfileMcpInfoResponse> {
        api::get_profile_mcp_info(
            &self.client.transport,
            &api::GetProfileMcpInfoParams {
                profile_id: self.profile_id.clone(),
            },
            self.client.options,
        )
        .await
    }

    pub async fn servers(&self) -> Result<models::McpServersListResponse> {
        api::list_profile_mcp_servers(
            &self.client.transport,
            &api::ListProfileMcpServersParams {
                profile_id: self.profile_id.clone(),
            },
            self.client.options,
        )
        .await
    }

    pub async fn default_permission(&self) -> Result<models::McpDefaultPermissionResponse> {
        api::get_profile_mcp_default(
            &self.client.transport,
            &api::GetProfileMcpDefaultParams {
                profile_id: self.profile_id.clone(),
            },
            self.client.options,
        )
        .await
    }

    pub async fn get(&self, name: &str) -> Result<ProfileMcpServer<'a>> {
        let matches = self
            .servers()
            .await?
            .0
            .into_iter()
            .filter(|server| server.name == name)
            .collect::<Vec<_>>();
        if matches.len() != 1 {
            return Err(crate::Error::InvalidInput("MCP server name must resolve exactly once"));
        }
        Ok(ProfileMcpServer {
            client: self.client,
            profile_id: self.profile_id.clone(),
            info: matches.into_iter().next().expect("one MCP server matched"),
        })
    }
}

impl<'a> ProfileMcpServer<'a> {
    pub fn tools(&self) -> McpTools<'a> {
        McpTools {
            client: self.client,
            profile_id: self.profile_id.clone(),
            server_id: self.info.name.clone(),
        }
    }

    pub async fn refresh(&self) -> Result<models::McpRefreshResponse> {
        api::refresh_profile_mcp_server(
            &self.client.transport,
            &api::RefreshProfileMcpServerParams {
                profile_id: self.profile_id.clone(),
                server_id: self.info.name.clone(),
            },
            self.client.options,
        )
        .await
    }
}

impl McpTools<'_> {
    pub async fn list(&self) -> Result<models::McpToolsListResponse> {
        api::list_profile_mcp_tools(
            &self.client.transport,
            &api::ListProfileMcpToolsParams {
                profile_id: self.profile_id.clone(),
                server_id: self.server_id.clone(),
            },
            self.client.options,
        )
        .await
    }

    pub async fn call(&self, name: &str, arguments: models::Value) -> Result<models::Value> {
        api::call_profile_mcp_tool(
            &self.client.transport,
            &api::CallProfileMcpToolParams {
                profile_id: self.profile_id.clone(),
                server_id: self.server_id.clone(),
                tool_id: name.into(),
                body: arguments,
            },
            self.client.options,
        )
        .await
    }
}
